src/main/java/htsjdk/samtools/util/GZIIndex.java

```diff
@@ -0,0 +1,365 @@
+ * uncompressed offset that that block starts with. This mapping is represented by {@link IndexEntry}.
+ *
+ * <p>An example of usage for this index for random access a bgzip file using an index generated
+ * from raw data. For example, for indexing a compressed FASTA file the .gzi index can be used in
+ * conjunction with the {@link htsjdk.samtools.reference.FastaSequenceIndex} to seek concrete
+ * sequences.
+ *
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ * @see <a href="http://github.com/samtools/htslib/issues/473">https://github.com/samtools/htslib/issues/473</a>
+ */
+public final class GZIIndex {
+
+    /** Default extension for the files storing a {@link GZIIndex}. */
+    public static final String DEFAULT_EXTENSION = ".gzi";
+
+    /**
+     * Index entry mapping the block-offset (compressed offset) to the uncompressed offset where the
+     * block starts.
+     *
+     * @see <a href="http://github.com/samtools/htslib/issues/473">https://github.com/samtools/htslib/issues/473</a>
+     */
+    public static final class IndexEntry {
+        private final long compressedOffset;
+        private final long uncompressedOffset;
+
+        private IndexEntry(final long compressedOffset, final long uncompressedOffset) {
+            if (Long.signum(compressedOffset) == -1) {
+                throw new IllegalArgumentException("negative compressed offset: " + compressedOffset);
+            }
+            if (Long.signum(uncompressedOffset) == -1) {
+                throw new IllegalArgumentException("negative uncompressed offset: " + uncompressedOffset);
+            }
+            this.compressedOffset = compressedOffset;
+            this.uncompressedOffset = uncompressedOffset;
+        }
+
+        /** Returns the compressed offset (block-offset). */
+        public long getCompressedOffset() {
+            return compressedOffset;
+        }
+
+        /** Returns the uncompressed offset. */
+        public long getUncompressedOffset() {
+            return uncompressedOffset;
+        }
+
+        @Override
+        public String toString() {
+            return String.format("IndexEntry={compressed=%d(0x%x),uncompressed=%d(0x%x)",
+                    compressedOffset, compressedOffset,
+                    uncompressedOffset, uncompressedOffset);
+        }
+
+        @Override
+        public boolean equals(final Object obj) {
+            if (obj == null || !(obj instanceof IndexEntry)) {
+                return false;
+            }
+            final IndexEntry other = (IndexEntry) obj;
+            return compressedOffset == other.compressedOffset
+                    && uncompressedOffset == other.uncompressedOffset;
+        }
+
+        @Override
+        public int hashCode() {
+            return 31 * Long.hashCode(compressedOffset) + Long.hashCode(uncompressedOffset);
+        }
+    }
+
+    // first entry in the mapping (do not write down)
+    private static final IndexEntry FIRST_MAPPING = new IndexEntry(0, 0);
```

Comment:
I wasn't sure if you needed this or not, I just saw it mentioned in the samtools issue.  If we don't need it for the query algorithm then we should just skip it.

Reply:
We do not really require this in the list; if there is no index entry where the uncompressed offset is larger than the requested one, then the first block is the one where the offset is (represented by this `FIRST_MAPPING`).